bothan-okx/src/service.rs

```diff
@@ -0,0 +1,205 @@
+) {
+    tokio::spawn(async move {
+        loop {
+            select! {
+                Some(cmd) = command_rx.recv() => {
+                    process_command(&cmd, &connection, &cache).await;
+                },
+                result = {
+                    let cloned = connection.clone();
+                    timeout(DEFAULT_TIMEOUT, async move { cloned.lock().await.next().await })
+                } => {
+                    let timestamp = chrono::Utc::now().timestamp();
+                    match &result {
+                        Ok(Ok(response)) => {
+                            process_response(response, &cache, timestamp as u64).await;
+                        },
+                        Ok(Err(OKXError::ChannelClosed)) | Err(_) => {
+                            // Attempt to reconnect on timeout or on channel close
+                            handle_reconnect(&connector, &connection, &cache, &command_tx).await;
+                        }
+                        Ok(Err(e)) => {
+                            error!("unexpected error: {}", e);
+                        }
+                    }
+                },
+                Some(ids) = removed_ids_rx.recv() => {
+                    let vec_ids = ids.iter().map(|x| x.as_str()).collect::<Vec<&str>>();
+                    if connection.lock().await.unsubscribe_ticker(vec_ids.as_slice()).await.is_err() {
+                        warn!("failed to unsubscribe to ids: {:?}", ids);
+                    }
+                }
+            }
+        }
+    });
+}
+
+async fn process_command(
+    cmd: &Command,
+    ws: &Mutex<OKXWebSocketConnection>,
+    cache: &Cache<PriceData>,
+) {
+    match cmd {
+        Command::Subscribe(ids) => {
+            cache.set_batch_pending(ids.clone()).await;
+
+            let vec_ids = ids.iter().map(|x| x.as_str()).collect::<Vec<&str>>();
+            let mut locked = ws.lock().await;
+            if locked.subscribe_ticker(vec_ids.as_slice()).await.is_err() {
+                warn!("Failed to subscribe to ids: {:?}", ids);
+            }
+        }
+    }
+}
+
+async fn handle_reconnect(
+    connector: &OKXWebSocketConnector,
+    connection: &Mutex<OKXWebSocketConnection>,
+    cache: &Cache<PriceData>,
+    command_tx: &Sender<Command>,
+) {
+    // TODO: Handle reconnection failure
+    let mut locked = connection.lock().await;
+    warn!("attempting to reconnect to okx");
+    // reconnect
+    if let Ok(new_connection) = connector.connect().await {
+        *locked = new_connection;
+    } else {
+        warn!("failed to reconnect to okx");
+        return;
+    }
+
+    //  resubscribe to all ids
+    let keys = cache.keys().await;
+
+    // If the keys are empty, subscribe to a single symbol to keep the connection alive
+    // TODO: find a better solution
+    let cmd = match keys.is_empty() {
+        true => Command::Subscribe(vec!["BTC-USDT".to_string()]),
+        false => Command::Subscribe(keys),
+    };
```

Comment:
does OKX also require a subscription to keep a connection like others?

Reply:
yes